### DISCUSSION TITLE
Better pick API with `useAsyncData` and `useFetch`

### DISCUSSION THREAD

**Discussion by pi0** [nuxt/nuxt#16831] [RFCs]
The idea originally proposed by @xtoolkit in #1389, we could allow picking nested keys from the response. (at the time of writing this, it is only possible to specify top-level keys for `pick`) 

Example response:

``` js
data = {
  id: 100,
  name: 'Alex',
  time: 111111111,
  info: {
    url: '/user/alex',
    like: {
      count: 165416,
      status: false
    },
    dislike: {
      count: 5456,
      status: false
    }
  }
}
```

Expected transformed data:

```js
user = {
  name: 'Alex',
  info: {
    url: '/user/alex',
    like: {
      status: false
    }
  }
}
```

We have  two options to achieve this:

- Improving `pick` API to support deep keys #1389
- Using `transform` option

I'm thinking that while the initial idea of introducing `pick` in #721 and improving in #1389 was an initiative, it will become soon opinionated since there is no single API method that covers all different use cases unless we do an overengineering to build an alternative of lodash directly inside `useFetch` or nuxt3 and bound to it. Alternatively, we could create a new unjs library to make a filtering API with better typing support and make it opt-in to use via the `transform` option.

(`createFilter` can be auto imported but tree shaken and added to bundle by usage)

```js
const { data: user } = await useFetch('/api/user/100', {
  // instead of pick: { /* options */ },
  transform: crateFilter(/* options */),
})
```

--- 

Appendix: Some ideas for extended `pick` API:

```js
// A: Nested array of string and object
const { data: user } = await useFetch('/api/user/100', {
  pick: [
    'name',
    {
      info: [
        'url',
        {
          like: ['status']
        }
      ]
    }
  ]
})

// B: GraphQL like object
const { data: user } = await useFetch('/api/user/100', {
  pick: {
    name: true, 
    info: {
      url: true,
      like: {
        status: true
      }
    }
  }
})

// C: Array with dotted keys
const { data: user } = await useFetch('/api/user/100', {
  pick: [ 'name', 'info.url', 'info.like.status' ]
})

// D: Nested arrays
const { data: user } = await useFetch('/api/user/100', {
  pick: [ 'name', ['info', ['url', ['like', 'status']]] ]
})
```

**Comment by Rigo-m** [comment]
I personally like the C option more. It's more concise and more readable

**Comment by wobsoriano** [comment]
C option reminds me of [feathers query](https://docs.feathersjs.com/api/databases/querying.html#select) which is readable

**Comment by chz** [comment]
Option C more convient

**Comment by jpsc** [comment]
Option C looks easier to understand without much context.

**Comment by xtoolkit** [comment]
i think B and C option is user friendly and Nuxt can use both.

**Comment by RonanOFM** [comment]
C option is more readable for me too.

But if you have to do this :

```
const { data: user } = await useFetch('/api/user/100', {
  pick: [ 
           'info.foo.status', 
           'info.foo.firstname', 
           'info.foo.lastname',
           'info.foo.email' ,
           ...
          ]
})
```
`info.foo` is repeated for each key, wich isn't very nice
However, i don't think this case often arise...( pick a lot of keys in deep levels )

**Comment by TheAlexLichter** [comment]
C is neat. Dot-notation is well-known (e.g. through lodash, feathers, mongo etc.) and easy to grasp 👍

**Comment by Rigo-m** [comment]
Any updates on this?

**Comment by sykes10** [comment]
I personally like the C option too. I use  GraphQL but I reckon that the B does not feel natural

**Comment by ragokan** [comment]
C looks the best but what about something like exclude?
For example:
```ts
const { data: user } = await useFetch('/api/user/100', {
  exclude: ["createdAt", "updatedAt", "someUnnecessaryData", "info.like.createdAt"]
})
```
So that we only declare required data and exclude what we don't want. If we want, we can alternatively use **Pick**, too.

**Comment by praburangki** [comment]
The C option is nice. But will it make harder to get the typings?

**Comment by Rigo-m** [comment]
One problem that I see with the pick in `useFetch`/`useAsyncData` is that it only works "client-side". It would be really cool to implement some way to pick data server side via `serverMiddleware` or a `/server/api` endpoint. This way, if I use a server api as some sort of a proxy/RPC to fetch data, I can pick data on the server and return a really slim response when the browser hits the api via CSR navigation.
I tried my best in the last couple of days to implement this but I guess I still sucks at typescript. @danielroe do you think this is feasible in any way?

**Comment by AxelBriche** [comment]
Hi, what about pick compatibility with a list result?

By example :

`const { data: users } = await useFetch('/api/users',  {pick: ...})`

**Reply by atinux** [reply]
I think we could smartly detect that if this is an array, we loop through it.

**Comment by ol1s** [comment]
Worth considering how this could work for an array of data:

```js
const { data: users } = await useFetch('/api/user', {
  pick: [ 
    '*.name', 
    '*.info.url', 
    '*.info.like.status' 
  ]
})
```

**Reply by atinux** [reply]
https://github.com/nuxt/nuxt/discussions/16831#discussioncomment-5381194

**Comment by AryeBorgen** [comment]
Why not use it like in TS interface like this example:
```ts
const { data: user } = await useFetch('/api/user/100',
pick :{
  user: {
    name: string,
    info: {
      url: string,
      like: {
        status: boolean,
      },
    },
  }
}

**Comment by NMTuan** [comment]
i like C 

and more:
`pick: ['id', 'name', 'info.url', 'info.like.*']`

**Comment by alphaCastor** [comment]
I like C as well. Looks clean and readable

**Comment by fmoessle** [comment]
Love C. Transform looks great too. It could be an alternative to Option C for more specific use-cases e.g. picking a lot of deep nested data.

**Comment by BracketJohn** [comment]
I  like option C but would suggest to replace the notation with [json pointer](https://www.rfc-editor.org/rfc/rfc6901):
```ts
// C.1: deep pick following the conventions to a T
const { data: user } = await useFetch('/api/user/100', {
  pick: [ '/name', '/info/url', '/info/like/status' ]
})

// C.2 deep pick without leading slash
const { data: user } = await useFetch('/api/user/100', {
  pick: [ 'name', 'info/url', 'info/like/status' ]
})
```

this has the advantages of being a broadly known notation + standard and the implementation can use one of the many existing packages (:

**Comment by Lionad-Morotar** [comment]
C is neat for simple usage
